dashboard/lib/expired_child_account_purger.rb

```diff
@@ -0,0 +1,212 @@
+    @log.puts "(dry-run)" if @dry_run
+  end
+
+  def check_constraints(accounts)
+    if accounts.count > @max_accounts_to_purge
+      @purge_size_limit_exceeded = 1
+      raise SafetyConstraintViolation, "Found #{accounts.count} " \
+        "accounts to purge, which exceeds the configured limit of " \
+        "#{@max_accounts_to_purge}. Abandoning run."
+    end
+  end
+
+  def report_results
+    review_queue_depth = QueuedAccountPurge.count
+    manual_reviews_needed = QueuedAccountPurge.needing_manual_review.count
+    metrics = build_metrics review_queue_depth, manual_reviews_needed
+    log_metrics metrics
+
+    summary = build_summary review_queue_depth, manual_reviews_needed
+    @log.puts summary
+
+    log_link = upload_activity_log
+    if rack_env? :production
+      if manual_reviews_needed > 0
+        warn "#{summary} #{log_link}"
+      else
+        say "#{summary} #{log_link}"
+      end
+    elsif rack_env? :development
+      puts "#{summary} #{log_link}"
+    end
+
+    upload_metrics metrics unless @dry_run
+  end
+
+  def build_metrics(review_queue_depth, manual_reviews_needed)
+    {
+      # 0 or 1 if the size of the purge would exceed the limit
+      PurgeSizeLimitExceeded: @purge_size_limit_exceeded,
+      # Number of accounts purged during this run
+      AccountsPurged: @num_accounts_purged,
+      # Number of accounts queued for manual review during this run
+      AccountsQueued: @num_accounts_queued,
+      # Depth of review queue after this run (may include auto-retryable entries)
+      ReviewQueueDepth: review_queue_depth,
+      # Total number of accounts needs manual review (including accounts from previous runs)
+      ManualReviewQueueDepth: manual_reviews_needed
+    }
+  end
+
+  def metric_name(name)
+    "Custom/ExpiredChildAccountPurger/#{name}"
+  end
+
+  def log_metrics(metrics)
+    metrics.each do |key, value|
+      @log.puts "#{key}: #{value}"
+    end
+  end
+
+  def upload_metrics(metrics)
+    aws_metrics = metrics.map do |key, value|
+      {
+        metric_name: key,
+        dimensions: [
+          {name: "Environment", value: CDO.rack_env},
+        ],
+        value: value
+      }
+    end
+    Cdo::Metrics.push('ExpiredChildAccountPurger', aws_metrics)
+  end
+
+  def build_summary(review_queue_depth, manual_reviews_needed)
+    formatted_duration = Time.at(Time.now.to_i - @start_time.to_i).utc.strftime("%H:%M:%S")
+
+    summary = expired_accounts_summary
+    summary += "\n" + purged_accounts_summary
+    summary + "\n🕐 #{formatted_duration}"
+  end
+
+  def purged_accounts_summary
+    intro = @dry_run ? 'Would have purged' : 'Purged'
+    "#{intro} #{@num_accounts_purged} child account(s)."
+  end
+
+  def expired_accounts_summary
+    "#{@num_accounts_to_be_purged} child account(s) have expired."
+  end
+
+  # @return [String] HTML link to view uploaded log
+  def upload_activity_log
+    log_url = AWS::S3::LogUploader.
```

Comment:
Even though we have a bunch of cron jobs that upload a batch of logs to S3 after completion, the preferred way to do it is to log along the way via `CDO.log`. [Example](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/email_reminder.rb#L41) from the email reminder cron job.

Reply:
I think this suggestion makes sense, however it would require me to make a bunch of changes to the shared class `AccountPurger` because it on works with IOStreams currently. I'm worried about making this PR longer than it already it is. Would it be okay if I created a JIRA task to do that as followup work?

Reply:
Sounds good - more important to ship it!